File: apps/guzo-api/src/main/java/et/guzo/service/SupportService.java
package et.guzo.service;

import et.guzo.common.ApiException;
import et.guzo.common.PaginationMeta;
import et.guzo.domain.entity.SupportTicket;
import et.guzo.domain.entity.TicketMessage;
import et.guzo.domain.entity.User;
import et.guzo.domain.enums.TicketPriority;
import et.guzo.domain.enums.TicketStatus;
import et.guzo.repository.SupportTicketRepository;
import et.guzo.repository.TicketMessageRepository;
import et.guzo.repository.UserRepository;
import et.guzo.security.AuthUser;
import et.guzo.security.RoleChecker;
import et.guzo.util.IdUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class SupportService {

    private final SupportTicketRepository ticketRepository;
    private final TicketMessageRepository messageRepository;
    private final UserRepository userRepository;

    private boolean isAgent(AuthUser user) {
        return RoleChecker.hasAnyRole(user, "SUPER_ADMIN", "ADMIN", "SUPPORT", "OPERATIONS_MANAGER");
    }

    @Transactional(readOnly = true)
    public Map<String, Object> list(AuthUser user, int page, int limit, String status, String priority, String search) {
        PageRequest pageable = PageRequest.of(Math.max(page - 1, 0), Math.max(limit, 1), Sort.by(Sort.Direction.DESC, "createdAt"));
        Page<SupportTicket> result;
        if (!isAgent(user)) {
            result = ticketRepository.findByRequesterId(user.getId(), pageable);
        } else if (search != null && !search.isBlank()) {
            result = ticketRepository.search(search, pageable);
        } else if (status != null && !status.isBlank()) {
            result = ticketRepository.findByStatus(TicketStatus.valueOf(status), pageable);
        } else {
            result = ticketRepository.findAll(pageable);
        }
        List<Map<String, Object>> items = result.getContent().stream().map(t -> toDto(t, user)).toList();
        return Map.of("items", items, "meta", PaginationMeta.of(page, limit, result.getTotalElements()));
    }

    @Transactional(readOnly = true)
    public Map<String, Object> get(String id, AuthUser user) {
        SupportTicket ticket = ticketRepository.findById(id).orElseThrow(() -> ApiException.notFound("Ticket not found"));
        assertAccess(ticket, user);
        return toDto(ticket, user, true);
    }

    @Transactional
    public Map<String, Object> create(Map<String, Object> body, AuthUser user) {
        Instant now = Instant.now();
        SupportTicket ticket = new SupportTicket();
        ticket.setId(IdUtil.cuid());
        ticket.setTicketNumber(IdUtil.ticketNumber());
        ticket.setRequesterId(user.getId());
        ticket.setSubject((String) body.get("subject"));
        ticket.setCategory((String) body.get("category"));
        if (body.get("priority") != null) ticket.setPriority(TicketPriority.valueOf(body.get("priority").toString()));
        if (body.get("orderId") != null) ticket.setOrderId(body.get("orderId").toString());
        ticket.setStatus(TicketStatus.OPEN);
        ticket.setCreatedAt(now);
        ticket.setUpdatedAt(now);
        ticketRepository.save(ticket);
        if (body.get("message") != null) {
            addMessage(ticket.getId(), Map.of("body", body.get("message")), user);
        }
        return toDto(ticket, user, true);
    }

    @Transactional
    public Map<String, Object> addMessage(String ticketId, Map<String, Object> body, AuthUser user) {
        SupportTicket ticket = ticketRepository.findById(ticketId).orElseThrow(() -> ApiException.notFound("Ticket not found"));
        assertAccess(ticket, user);
        boolean internal = isAgent(user) && Boolean.TRUE.equals(body.get("isInternal"));
        Instant now = Instant.now();
        TicketMessage msg = new TicketMessage();
        msg.setId(IdUtil.cuid());
        msg.setTicketId(ticketId);
        msg.setAuthorId(user.getId());
        msg.setBody(body.get("body").toString());
        msg.setInternal(internal);
        msg.setCreatedAt(now);
        messageRepository.save(msg);
        if (!internal) {
            if (isAgent(user) && ticket.getStatus() == TicketStatus.OPEN) ticket.setStatus(TicketStatus.IN_PROGRESS);
            if (!isAgent(user) && List.of(TicketStatus.RESOLVED, TicketStatus.CLOSED, TicketStatus.WAITING_CUSTOMER).contains(ticket.getStatus())) {
                ticket.setStatus(TicketStatus.OPEN);
            }
            ticket.setUpdatedAt(now);
            ticketRepository.save(ticket);
        }
        return Map.of("id", msg.getId(), "body", msg.getBody(), "isInternal", msg.isInternal(), "createdAt", msg.getCreatedAt(),
            "author", person(user.getId()));
    }

    @Transactional
    public Map<String, Object> update(String id, Map<String, Object> body, AuthUser user) {
        if (!isAgent(user)) throw ApiException.forbidden("Agent access required");
        SupportTicket ticket = ticketRepository.findById(id).orElseThrow(() -> ApiException.notFound("Ticket not found"));
        Instant now = Instant.now();
        if (body.get("status") != null) {
            TicketStatus status = TicketStatus.valueOf(body.get("status").toString());
            ticket.setStatus(status);
            ticket.setResolvedAt(status == TicketStatus.RESOLVED || status == TicketStatus.CLOSED ? now : null);
        }
        if (body.get("priority") != null) ticket.setPriority(TicketPriority.valueOf(body.get("priority").toString()));
        if (body.containsKey("assigneeId")) {
            ticket.setAssigneeId(body.get("assigneeId") != null ? body.get("assigneeId").toString() : null);
        }
        if (body.get("category") != null) ticket.setCategory(body.get("category").toString());
        ticket.setUpdatedAt(now);
        ticketRepository.save(ticket);
        return toDto(ticket, user, true);
    }

    private void assertAccess(SupportTicket ticket, AuthUser user) {
        if (!isAgent(user) && !ticket.getRequesterId().equals(user.getId())) {
            throw ApiException.forbidden("Not your ticket");
        }
    }

    private Map<String, Object> toDto(SupportTicket t, AuthUser user) {
        return toDto(t, user, false);
    }

    private Map<String, Object> toDto(SupportTicket t, AuthUser user, boolean withMessages) {
        List<TicketMessage> messages = withMessages ? messageRepository.findByTicketIdOrderByCreatedAtAsc(t.getId()) : List.of();
        List<Map<String, Object>> msgDtos = messages.stream()
            .filter(m -> isAgent(user) || !m.isInternal())
            .map(m -> Map.<String, Object>of(
                "id", m.getId(), "body", m.getBody(), "isInternal", m.isInternal(),
                "createdAt", m.getCreatedAt(), "author", person(m.getAuthorId())
            )).toList();
        Map<String, Object> m = new java.util.HashMap<>();
        m.put("id", t.getId());
        m.put("ticketNumber", t.getTicketNumber());
        m.put("subject", t.getSubject());
        m.put("category", t.getCategory());
        m.put("status", t.getStatus().name());
        m.put("priority", t.getPriority().name());
        m.put("orderId", t.getOrderId());
        m.put("createdAt", t.getCreatedAt());
        m.put("updatedAt", t.getUpdatedAt());
        m.put("resolvedAt", t.getResolvedAt());
        m.put("requester", person(t.getRequesterId()));
        m.put("assignee", t.getAssigneeId() != null ? person(t.getAssigneeId()) : null);
        m.put("messages", msgDtos);
        return m;
    }

    private Map<String, Object> person(String userId) {
        User u = userRepository.findById(userId).orElse(null);
        if (u == null) return Map.of("id", userId, "firstName", "Unknown", "lastName", "");
        return Map.of("id", u.getId(), "email", u.getEmail(), "firstName", u.getFirstName(), "lastName", u.getLastName());
    }
}

File: apps/server/src/modules/support/support.service.ts
import { supportRepository, SupportTicketRepository } from './support.repository.js';
import { ApiError } from '../../utils/ApiError.js';
import { buildMeta, type PaginationMeta } from '../../utils/ApiResponse.js';
import type { ParsedListQuery } from '../../utils/pagination.js';
import type { CreateSupportTicketDto, UpdateSupportTicketDto, AddTicketMessageDto } from './support.dto.js';
import { generateReference } from '@delivery/utils';
import { emitToUser } from '../../socket/index.js';
import { SOCKET_EVENTS } from '@delivery/types';

export interface TicketContext {
  userId: string;
  isAgent: boolean;
}

interface TicketRow {
  requesterId: string;
  assigneeId: string | null;
  ticketNumber: string;
  status: string;
}

export class SupportTicketService {
  constructor(private readonly repo: SupportTicketRepository = supportRepository) {}

  async list(
    query: ParsedListQuery,
    scope?: { requesterId?: string },
  ): Promise<{ items: unknown[]; meta: PaginationMeta }> {
    const { items, total } = await this.repo.list({
      skip: query.skip,
      take: query.take,
      search: query.search,
      status: query.filters.status,
      priority: query.filters.priority,
      assigneeId: query.filters.assigneeId,
      requesterId: scope?.requesterId,
      sortBy: query.sortBy,
      sortOrder: query.sortOrder,
    });
    return { items, meta: buildMeta(query.page, query.limit, total) };
  }

  async getById(id: string, ctx: TicketContext) {
    const ticket = await this.repo.findById(id);
    if (!ticket) throw ApiError.notFound('Support ticket not found');
    this.assertAccess(ticket, ctx);
    if (!ctx.isAgent) {
      return { ...ticket, messages: ticket.messages.filter((m) => !m.isInternal) };
    }
    return ticket;
  }

  async create(dto: CreateSupportTicketDto, userId: string) {
    const ticketNumber = generateReference('TIC');
    const ticket = await this.repo.create({
      ticketNumber,
      subject: dto.subject,
      category: dto.category,
      priority: dto.priority ?? 'MEDIUM',
      orderId: dto.orderId,
      requester: { connect: { id: userId } },
      ...(dto.message
        ? { messages: { create: [{ body: dto.message, author: { connect: { id: userId } } }] } }
        : {}),
    });
    return ticket;
  }

  async addMessage(id: string, dto: AddTicketMessageDto, ctx: TicketContext) {
    const ticket = (await this.repo.findById(id)) as (TicketRow & { id: string }) | null;
    if (!ticket) throw ApiError.notFound('Support ticket not found');
    this.assertAccess(ticket, ctx);

    const isInternal = ctx.isAgent ? !!dto.isInternal : false;
    const message = await this.repo.addMessage({
      ticketId: id,
      authorId: ctx.userId,
      body: dto.body,
      isInternal,
    });

    if (!isInternal) {
      let nextStatus: string | undefined;
      if (ctx.isAgent && ticket.status === 'OPEN') nextStatus = 'IN_PROGRESS';
      if (!ctx.isAgent && ['RESOLVED', 'CLOSED', 'WAITING_CUSTOMER'].includes(ticket.status)) nextStatus = 'OPEN';
      await this.repo.touch(id, nextStatus ? { status: nextStatus as never } : {});

      const notifyUserId = ctx.isAgent ? ticket.requesterId : ticket.assigneeId;
      if (notifyUserId) {
        emitToUser(notifyUserId, SOCKET_EVENTS.NOTIFICATION_NEW, {
          type: 'SUPPORT_REPLY',
          title: `Ticket ${ticket.ticketNumber}`,
          body: ctx.isAgent ? 'Support replied to your ticket' : 'Customer replied to a ticket',
        });
      }
    }

    return message;
  }

  async update(id: string, dto: UpdateSupportTicketDto, ctx: TicketContext) {
    const ticket = (await this.repo.findById(id)) as (TicketRow & { id: string }) | null;
    if (!ticket) throw ApiError.notFound('Support ticket not found');

    const isResolving = dto.status === 'RESOLVED' || dto.status === 'CLOSED';
    const updated = await this.repo.update(id, {
      ...(dto.status ? { status: dto.status as never } : {}),
      ...(dto.priority ? { priority: dto.priority as never } : {}),
      ...(dto.category !== undefined ? { category: dto.category } : {}),
      ...(dto.assigneeId !== undefined
        ? dto.assigneeId
          ? { assignee: { connect: { id: dto.assigneeId } } }
          : { assignee: { disconnect: true } }
        : {}),
      ...(isResolving ? { resolvedAt: new Date() } : {}),
      ...(dto.status && !isResolving ? { resolvedAt: null } : {}),
    });

    if (dto.status && ticket.requesterId !== ctx.userId) {
      emitToUser(ticket.requesterId, SOCKET_EVENTS.NOTIFICATION_NEW, {
        type: 'SUPPORT_STATUS',
        title: `Ticket ${ticket.ticketNumber}`,
        body: `Your ticket is now ${dto.status}`,
      });
    }
    return updated;
  }

  async remove(id: string): Promise<void> {
    const ticket = await this.repo.findById(id);
    if (!ticket) throw ApiError.notFound('Support ticket not found');
    await this.repo.delete(id);
  }

  private assertAccess(ticket: { requesterId: string }, ctx: TicketContext) {
    if (!ctx.isAgent && ticket.requesterId !== ctx.userId) {
      throw ApiError.forbidden('You do not have access to this ticket');
    }
  }
}

export const supportService = new SupportTicketService();
